pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -0,0 +1,1134 @@
+    BrokerSubscription, BrokerUpdate, Client,
+};
+use tokio::select;
+use tracing::*;
+
+use crate::DatadirTimelineImpl;
+use utils::{
+    lsn::Lsn,
+    pq_proto::ReplicationFeedback,
+    zid::{NodeId, ZTenantTimelineId},
+};
+
+use super::{TaskEvent, TaskHandle};
+
+/// Spawns the loop to take care of the timeline's WAL streaming connection.
+pub(super) fn spawn_connection_manager_task(
+    id: ZTenantTimelineId,
+    broker_loop_prefix: String,
+    mut client: Client,
+    local_timeline: Arc<DatadirTimelineImpl>,
+    wal_connect_timeout: Duration,
+    lagging_wal_timeout: Duration,
+    max_lsn_wal_lag: NonZeroU64,
+) -> TaskHandle<()> {
+    TaskHandle::spawn(move |_, mut cancellation| {
+        async move {
+            info!("WAL receiver broker started, connecting to etcd");
+            let mut walreceiver_state = WalreceiverState::new(
+                id,
+                local_timeline,
+                wal_connect_timeout,
+                lagging_wal_timeout,
+                max_lsn_wal_lag,
+            );
+            loop {
+                select! {
+                    _ = cancellation.changed() => {
+                        info!("Broker subscription init cancelled, shutting down");
+                        if let Some(wal_connection) = walreceiver_state.wal_connection.take()
+                        {
+                            wal_connection.connection_task.shutdown().await;
+                        }
+                        return Ok(());
+                    },
+
+                    _ = connection_manager_loop_step(
+                        &broker_loop_prefix,
+                        &mut client,
+                        &mut walreceiver_state,
+                    ) => {},
+                }
+            }
+        }
+        .instrument(info_span!("wal_connection_manager", id = %id))
+    })
+}
+
+/// Attempts to subscribe for timeline updates, pushed by safekeepers into the broker.
+/// Based on the updates, desides whether to start, keep or stop a WAL receiver task.
+/// If etcd subscription is cancelled, exits.
+async fn connection_manager_loop_step(
+    broker_prefix: &str,
+    etcd_client: &mut Client,
+    walreceiver_state: &mut WalreceiverState,
+) {
+    let id = walreceiver_state.id;
+
+    let mut broker_subscription =
+        subscribe_for_timeline_updates(etcd_client, broker_prefix, id).await;
+    info!("Subscribed for etcd timeline changes, waiting for new etcd data");
+
+    loop {
+        select! {
+            // the order of the polls is especially important here, since the first task to complete gets selected and the others get dropped (cancelled).
+            // place more frequetly updated tasks below to ensure the "slow" tasks are also reacted to.
+            biased;
+
+            // First, check if the broker connection is intact, otherwise we need to restart the whole step
+            broker_connection_result = &mut broker_subscription.watcher_handle => {
```

Comment:
Since closing watch channel and task ending don't happen atomically, there is a chance of getting out through `broker_subscription.value_updates.recv()` arm `returning None`, in which case again error is swallowed.